### PR TITLE
generalize container episode

### PR DESCRIPTION
**Pull request by wikfeldt**
biggest change: introduce conda as a package and environment manager.
TODO:
- describe how to create and share your own conda package
- see if conda section can be shortened
- introduce conda exercise
- reduce length of docker material

### CODE REVIEW

**Comment by Sabryr** [comment]
Looks good so far

**Comment by bast** [comment]
I like the pros/cons suggestion! Please also somewhere mention pipenv (perhaps that is already the case).
Also miniconda vs. conda, conda vs. pip, virtualenv vs. pipenv.

**Comment by wikfeldt** [comment]
ready for review and merging

File: _episodes/04-reproducible-envs.md [hunk 1]
```diff
@@ -244,34 +254,66 @@ $ conda activate word-count
...
 
+- When it comes to packaging and dependency management in Python 
+  (which can be rather convoluted and hard to understand) many alternative 
+  tools exist.
 
+| Tool | Purpose | Comments |
+| ----- | ------ | -------- |
+| [pip](https://pypi.org/project/pip/) | Python package installer | Can be used with conda. | 
+| [virtualenv](https://virtualenv.pypa.io/en/latest/) | Tool to create isolated Python environments | Partly integrated into standard library under `venv` module. |
+| [pipenv](https://pipenv.readthedocs.io/en/latest/) | Python package and virtualenv management | Official PyPA recommendation, replaces `pip` and `virtualenv`. |
```
Comment (bast) [inline _episodes/04-reproducible-envs.md]: I am not sure it is really a replacement. One could get the impression that `pip` and `venv` are now obsolete. I would rather say that it integrates them better and provides a higher level interface.

**Comment by bast** [comment]
Under containers please also mention that it can be a way of "sending the computer to the data" if sending data is not possible (too much data or data is sensitive).

Also we had the discussion thread on the chat about containers and some of this could go into a pros/cons. Like: containers as a way to not write portable software, on the other hand possibly the only way to make software reproducible which has external dependencies. But this is not a merge blocker.

**Comment by wikfeldt** [comment]
thanks for input. I already added a bullet about "sending computer to data if data big/sensitive", but i should add about the pros/cons of containers. I'll also clarify the purpose of pipenv

**Comment by bast** [comment]
Great! Sorry I missed that one point when browsing.

**Comment by bast** [comment]
And in that pros/cons discussion I had some opinions and of course I am not asking to advocate for these opinions as truth - but I thought having 2-3 sentences on pros/cons could spark interesting discussions with participants.

**Comment by wikfeldt** [comment]
i've now added a section on pros/cons and a discussion exercise on reproducibility aspects. And clarified role of pipenv.

**Comment by wikfeldt** [comment]
merging myself after two approvals

**Pull request merged**